snapcraft/plugins/qbs.py

```diff
@@ -0,0 +1,146 @@
+      will be used.
+    - profile:
+      (enum, 'gcc' or 'clang')
+      Set either gcc or clang as the base build profile
+"""
+
+import os
+
+import snapcraft
+
+
+class QbsPlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema['properties']['options'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        schema['properties']['profile'] = {
+            'enum': ['gcc', 'clang'],
+            'default': 'gcc',
+        }
+
+        schema['properties']['build-variant'] = {
+            'enum': ['debug', 'release'],
+            'default': 'release',
+        }
+
+        schema['properties']['qt-version'] = {
+            'enum': ['nil', 'qt4', 'qt5'],
+        }
+
+        schema['required'].append('qt-version')
+
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the pull step dirty.
+        return ['qt-version', 'profile']
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['options', 'build-variant']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['qbs', 'gcc', 'clang'])
+
+        if self.options.qt_version == 'qt5':
+            self.build_packages.extend(['qt5-qmake', 'qtbase5-dev'])
+        elif self.options.qt_version == 'qt4':
+            self.build_packages.extend(['qt4-qmake', 'libqt4-dev'])
+        elif self.options.qt_version != 'nil':
+            raise RuntimeError('Unsupported Qt version: {!r}'.format(
+                self.options.qt_version))
+
+    def build(self):
+        super().build()
+
+        env = self._build_environment()
+
+        # Setup the toolchains, there will only be gcc or clang by default
+        self.run(['qbs', 'setup-toolchains', '--detect'], env=env)
+
+        build_profile = ""
+
+        if self.options.qt_version == 'nil':
+            build_profile = self.options.profile
+        else:
+            # a unique'ish name for snap builds. Hopefully this shouldn't clash
+            # with any other local profiles. Each profile should look something
+            # like: profiles.snapcraftqt5clang or profiles.snapcraftnilgcc
+            build_profile = 'snapcraft' + \
+                            self.options.qt_version + \
+                            self.options.profile
+
+            # Setup the Qt profile.
+            self.run(['qbs', 'setup-qt', '/usr/bin/qmake', build_profile],
```

Comment:
since calling qmake on a specific path, shouldn't you before do this:

```
if not os.path.exists('/usr/bin/qmake'):
    raise RuntimeError('The qbs plugin requires qmake on /usr/bin/qmake')
```

Do you also want to hardcode it to that path?